Library/Homebrew/migrator.rb

```
@@ -0,0 +1,327 @@
+    # is the same as for core formula.
+    if (path = old_tabs.first.source["path"]) \
+        && path != Formulary.core_path(formula.oldname).to_s
+      old_core = false
+    end
+    new_core && old_core
+  end
+
+  # Fix INSTALL_RECEIPTS for tap-migrated formula.
+  def fix_tabs
+    old_tabs.each do |tab|
+      tab.source["tap"] = formula.tap
+      tab.write
+    end
+  end
+
+  def from_same_taps?
+    if both_from_core?
+      true
+    elsif formula.tap == old_tap
+      true
+    # Homebrew didn't use to update tabs while performing tap-migrations,
+    # so there can be INSTALL_RECEIPT's containing wrong information about
+    # tap (tap is Homebrew/homebrew if installed formula migrates to a tap), so
+    # we check if there is an entry about oldname migrated to tap and if
+    # newname's tap is the same as tap to which oldname migrated, then we
+    # can perform migrations and the taps for oldname and newname are the same.
+    else
+      require "tap_migrations"
+      if TAP_MIGRATIONS && (rec = TAP_MIGRATIONS[formula.oldname]) \
+          && rec.sub("/", "/homebrew-") == formula.tap
+        fix_tabs
+        true
+      else
+        false
+      end
+    end
+  end
+
+  def get_linked_oldkeg
+    kegs = oldpath.subdirs.map { |d| Keg.new(d) }
+    kegs.detect(&:linked?) || kegs.detect(&:optlinked?)
+  end
+
+  def pinned?
+    @pinned
+  end
+
+  def oldkeg_linked?
+    !!oldkeg
+  end
+
+  def migrate
+    if newpath.exist?
+      opoo "#{newpath} already exists; remove it manually and run brew migrate #{oldname}."
+    else
+      begin
+        oh1 "Migrating #{Tty.green}#{oldname}#{Tty.white} to #{Tty.green}#{newname}#{Tty.reset}"
+        copy_to_new_directory
+        repin
+        remove_and_unlink_oldname
+        link_newname
+        link_oldname_opt
+        link_oldname_cellar
+        update_tabs
+      # TODO
+      # - Make some output about exception.
+      rescue Exception
```

Although another option would be to silently catch it, say you should rerun `brew migrate foo` and then we print the exception out from that.